samples/net/mqtt_publisher/src/main.c

```diff
@@ -259,6 +269,9 @@ static void client_init(struct mqtt_client *client)
 
 	/* MQTT client configuration */
 	client->broker = &broker;
+#if defined(CONFIG_SOCKS) && !defined(CONFIG_MQTT_LIB_TLS)
```

Comment:
@rlubos Thank you for all your comments.

It cannot really be done with runtime checks here, because some of the fields in the header are already `ifdef`'d - so we have to either include them at all times, or use the compile-time macros here too.
See:
https://github.com/zephyrproject-rtos/zephyr/blob/bb50d75fc4ff146b38a25117b76f9bc720776bbb/include/net/mqtt.h#L372-L383

Would leaving `#if defined(...)` here be ok? Or what do you suggest?

Reply:
Ah I see the problem...

@tbursztyka How de we approach such issue, when some struct fields are conditionally included? Shall we remove `#if defined(...)` from the struct definition, or stick to `#if defined(...)` in code?

Reply:
@rlubos 
There is also an issue with this:
https://github.com/zephyrproject-rtos/zephyr/blob/bb50d75fc4ff146b38a25117b76f9bc720776bbb/include/net/mqtt.h#L341-L355

If we add another transport, and explicitly assign it the next number (`0x02`), and then add the new definition here:
https://github.com/zephyrproject-rtos/zephyr/blob/bb50d75fc4ff146b38a25117b76f9bc720776bbb/subsys/net/lib/mqtt/mqtt_transport.c#L33-L48

And then say compile it with the new transport (`0x02`) enabled but the transport in between disabled (`CONFIG_MQTT_LIB_TLS` - `0x01`), then this will fail:
https://github.com/zephyrproject-rtos/zephyr/blob/bb50d75fc4ff146b38a25117b76f9bc720776bbb/subsys/net/lib/mqtt/mqtt_transport.c#L52

Because the enum value will no longer be equal to the array index.

So we have to either stop assigning the numbers in the enum explicitly, or handle the `#if..` differently here too.

Reply:
Looks like getting rid `#if defined(CONFIG_MQTT_LIB_TLS)` is quite cumbersome in this case. Then I think we could stick to `#if defined(...)` nomenclature where needed, just as @tbursztyka commented earlier. 

Getting rid of explicit enum numeration sound good to me - currently it's kinda messy anyway (like fixed number of transports even though it's actually configurable).